hitless/pool_hook.go

```diff
@@ -0,0 +1,477 @@
+		handoffQueue: make(chan HandoffRequest, config.HandoffQueueSize),
+		// shutdown is a channel for signaling shutdown
+		shutdown:      make(chan struct{}),
+		maxWorkers:    config.MaxWorkers,
+		activeWorkers: 0,                // Start with no workers - create on demand
+		workerTimeout: 30 * time.Second, // Workers exit after 30s of inactivity
+		config:        config,
+		// Hitless manager for operation completion tracking
+		hitlessManager: hitlessManager,
+	}
+
+	// No upfront worker creation - workers are created on demand
+
+	return ph
+}
+
+// SetPool sets the pool interface for removing connections on handoff failure
+func (ph *PoolHook) SetPool(pooler pool.Pooler) {
+	ph.pool = pooler
+}
+
+// GetCurrentWorkers returns the current number of active workers (for testing)
+func (ph *PoolHook) GetCurrentWorkers() int {
+	return int(atomic.LoadInt32(&ph.activeWorkers))
+}
+
+// GetScaleLevel returns 1 if workers are active, 0 if none (for testing compatibility)
+func (ph *PoolHook) GetScaleLevel() int {
+	if atomic.LoadInt32(&ph.activeWorkers) > 0 {
+		return 1
+	}
+	return 0
+}
+
+// IsHandoffPending returns true if the given connection has a pending handoff
+func (ph *PoolHook) IsHandoffPending(conn *pool.Conn) bool {
+	_, pending := ph.pending.Load(conn.GetID())
+	return pending
+}
+
+// OnGet is called when a connection is retrieved from the pool
+func (ph *PoolHook) OnGet(ctx context.Context, conn *pool.Conn, isNewConn bool) error {
+	// NOTE: There are two conditions to make sure we don't return a connection that should be handed off or is
+	// in a handoff state at the moment.
+
+	// Check if connection is usable (not in a handoff state)
+	// Should not happen since the pool will not return a connection that is not usable.
+	if !conn.IsUsable() {
+		return ErrConnectionMarkedForHandoff
+	}
+
+	// Check if connection is marked for handoff, which means it will be queued for handoff on put.
+	if conn.ShouldHandoff() {
+		return ErrConnectionMarkedForHandoff
+	}
+
+	return nil
+}
+
+// OnPut is called when a connection is returned to the pool
+func (ph *PoolHook) OnPut(ctx context.Context, conn *pool.Conn) (shouldPool bool, shouldRemove bool, err error) {
+	// first check if we should handoff for faster rejection
+	if conn.ShouldHandoff() {
+		// check pending handoff to not queue the same connection twice
+		_, hasPendingHandoff := ph.pending.Load(conn.GetID())
+		if !hasPendingHandoff {
+			// Check for empty endpoint first (synchronous check)
+			if conn.GetHandoffEndpoint() == "" {
+				conn.ClearHandoffState()
+			} else {
+				if err := ph.queueHandoff(conn); err != nil {
+					// Failed to queue handoff, remove the connection
+					internal.Logger.Printf(ctx, "Failed to queue handoff: %v", err)
+					return false, true, nil // Don't pool, remove connection, no error to caller
+				}
+
+				// Check if handoff was already processed by a worker before we can mark it as queued
+				if !conn.ShouldHandoff() {
```

Comment:
There’s a TOCTOU race in 155–170: OnPut() checks conn.ShouldHandoff(), but a worker can finish the handoff and call conn.ClearHandoffState() before OnPut() marks it, causing conn.MarkQueuedForHandoff() to fail. 
Maybe for safety replace the check+mark sequence with a single atomic TryMarkQueuedForHandoff() using compare and swap to close the race and avoid connection loss under load.